Deprecate the usage of :class_name for :through association (#35830)
rails/rails: activerecord/CHANGELOG.md

*   Deprecate `:class_name` option for the `:through` association

>     Through association should always relay on the :class_name of its source

Reviewer: ```suggestion
    Through association should always relay on the `:class_name` of its source
```

>     Through association should always relay on the `:class_name` of its source

Reviewer 2: ```suggestion
    A through association should always rely on the `:class_name` of its source
```

    association. Example:

    ```
    Post.has_many :commenters, through: :comments,
      source: :author, class_name: 'CommentAuthor'
    post.comments.first.author.class # => Author
    post.commenters.first.class # => CommentAuthor
    ```

>     That causes inconsistency to the through association general sense and

Reviewer 2: I'm not sure that I fully understand this statement. Specifically, I'm not sure I understand what is meant by ". . .that assumes that principle to be true." I made a couple of rewording suggestions:
```suggestion
    Using `:class_name` causes inconsistencies to the through association including
    some bugs in preloading.
```

Author: The principle is (by my example in the description): `post.commenters == post.flat_map(&:comments).map(&:author)` when the association has no custom scope. Please help me to describe that.

*   Deprecate `:class_name` option for the `:through` association

>     Through association should always relay on the :class_name of its source

Reviewer 2: Spelling correction:
```suggestion
    Through association should always rely on the :class_name of its source
```

Reviewer: If we can quote the options here that would be more readable I think. 

```suggestion
    The `:through` association should always rely on the `:class_name` of its source.
```
